24-25-Server-Assignment-04_김태우

TwooTwoo opened the pull request:
## ERD
![image](https://github.com/user-attachments/assets/b8478cd8-df3c-4098-84dd-865bc93f5414)

## Description
<!-- 구현 및 작업 내용을 적어주세요 -->

### 1) Course

| Method | Endpoint                  | Description     |
|--------|----------------------------|-----------------|
| POST   | /courses                         | 강의 생성       |
| GET    | /courses/{courseId}         | 강의 조회       |
| GET    | /courses/                         | 전체 강의 조회 |
| PATCH  | /courses/{courseId}        | 강의 수정       |
| DELETE | /courses/{courseId}        | 강의 삭제       |

### 2) Student

| Method | Endpoint                  | Description     |
|--------|----------------------------|-----------------|
| POST   | /students                       | 학생 생성       |
| GET    | /students/{studentId}      | 학생 조회       |
| GET    | /students                         | 전체 학생 조회 |
| PATCH  | /students/{studentId}      | 학생 수정       |
| DELETE | /students/{studentId}      | 학생 삭제       |

### 3) Sugang

| Method | Endpoint                                      | Description               |
|--------|---------------------------------------|---------------------------|
| POST   | /courseRegistration                                        | 수강신청 등록            |
| GET    | /courseRegistration/{courseRegistrationId}                | 수강신청 조회       |
| GET    | /courseRegistration/findAll               | 전체 수강신청 조회      |
| DELETE | /courseRegistration/{courseRegistrationId} | 수강신청 취소      |
| GET    | /courseRegistration/findAll/{studentNumber}                  | 학번을 통해 해당 학생의 수강신청 내역 조회 |


## Important content

<!-- 주의 깊게 봐줬으면 하는 부분을 적어주세요 -->

- 수정 및 재제출합니다. 늦어서 죄송합니다.
- 모르거나 헷갈리는 부분, 공부한 개념을 주석으로 적어놓았습니다. 더 깊게 파고 들어갈 수 있는 부분을 많이 알려주시면 좋을 것 같습니다!

## Question

<!-- 궁금한 점을 적어주세요 -->
- 정적 팩토리 메서드에 대해 많이 찾아 보았습니다. of 와 from 등의 작명 규칙을 따를 수도 있고, 필요에 따라 적절한 이름을 작명할 수도 있는 것으로 알고 있습니다. 실무에서는 이 두가지 방식 중 어떤 것을 주로 사용하는지 궁금합니다.

## Reference
<!-- 참고한 레퍼런스가 있다면 공유해 주세요 -->
* 많이 헤멘 오류. 아직도 완벽하게 이해하지 못했습니다.
    + https://cantcoding.tistory.com/78
* 다른 멤버들의 PR을 많이 참고했습니다.

TwooTwoo committed:
김태우 4주차 과제 미완성
수강신청 컨트롤러 코드를 작성하지 못했습니다

TwooTwoo committed:
Fix : 오타 및 수강신청 레포지토리 객체의 메서드 이름 수정

TwooTwoo committed:
Fix : 코드 작성한 부분까지 포스트맨 테스트 결과 추가

TwooTwoo committed:
학생 CRUD 기능 구현 완료 및 포스트맨 테스트 결과 업로드

TwooTwoo committed:
Feat : 수강신청 기능 구현 완료 및 포스트맨 테스트 결과 업로드

TwooTwoo committed:
Refactor : ExceptionController 클래스 추가로 예외 처리 도입
그 외 불필요한 코드 삭제 등 일부 수정

TwooTwoo committed:
Chore : 가독성 향상 작업
일관적인 개행 규칙 적용
Service 계층의 메서드 순서 변경
일관적인 어노테이션 순서 적용
그 외 가독형 향상을 위한 자잘한 코드 수정

TwooTwoo committed:
Chore : 포스트맨 사진 제목 수정, ExceptionController 패키지 이동
역할의 분리를 위해 ExceptionController 객체를 controller 패키지에서 exception 패키지로 이동했습니다.

TwooTwoo committed:
Chore : 주석 오타 수정

hangillee commented on KIMTAEWOO/GDGWeek4AssignmentKTW/src/main/java/org/example/gdgweek4assignmentktw/controller/CourseController.java:
```diff
@@ -0,0 +1,52 @@
+        return new ResponseEntity<>(deleteResponseMessage, HttpStatus.OK);
+    }
+    
+}
```
[POSIX New Line](https://hyeon9mak.github.io/github-no-newline-at-a-end-of-file/)에 대해 학습해봅시다!

TwooTwoo replied:
POSIX에 이런게 있었군요. 인텔리제이에서 자동 설정도 가능하네요. 이런 기본적인 걸 챙기지 못하고 있었던 게 부끄럽습니다. 알려주셔서 감사합니다!

hangillee commented on KIMTAEWOO/GDGWeek4AssignmentKTW/src/main/java/org/example/gdgweek4assignmentktw/controller/CourseController.java:
```diff
@@ -0,0 +1,52 @@
+
+    // courseId 로 강의 수정m      
+    @PatchMapping("/{courseId}")
+    public ResponseEntity<CourseInfoResponseDto> updateByCourseId(@PathVariable Long courseId
+            , @RequestBody CourseSaveRequestDto dto) {
```
개행이 어색한 것 같습니다.. 쉼표가 왜 갑자기 나오게 되었을까요..?

TwooTwoo replied:
IDE의 자동 줄바꿈을 설정하지 않고 사용하고 있었습니다. 그러다 보니 무의식적으로 가로로 긴 코드 중간에서 줄바꿈을 했습니다. 인텔리제이의 soft-wrap 옵션을 활성화했으며, 앞으로 개행에 더 신경쓰도록 하겠습니다.

hangillee commented on KIMTAEWOO/GDGWeek4AssignmentKTW/src/main/java/org/example/gdgweek4assignmentktw/controller/CourseRegistrationController.java:
```diff
@@ -0,0 +1,58 @@
+@RequestMapping("/courseRegistration")
+public class CourseRegistrationController {
+
+    private final CourseRegistrationService courseRegistrationService;
+    private final CourseRepository courseRepository;
+    private final StudentRepository studentRepository;
```
컨트롤러 계층에서 `Service`와 `Repository` 계층 둘 모두를 의존하게 있는게 좋은 설계인 것 같진 않은 것 같습니다.

특히, 비즈니스 로직이 관례상 서비스 계층에 있게 될텐데, 비즈니스 로직에서 관리되지 않는 영속화 과정이 컨트롤러 계층에서 작성될 위험이 열리게 되는 것 같습니다.

TwooTwoo replied:
맞는 말씀입니다. 코드를 다시 확인해 꼭 필요한 서비스 객체의 의존성만 주입하도록 변경하여 다시 커밋하도록 하겠습니다.

hangillee commented on KIMTAEWOO/GDGWeek4AssignmentKTW/src/main/java/org/example/gdgweek4assignmentktw/controller/CourseRegistrationController.java:
```diff
@@ -0,0 +1,58 @@
+    // courseRegistrationId 로 수강신청 조회
+    @GetMapping("/{courseRegistrationId}")
+    public ResponseEntity<CourseRegistrationResponseDto> checkRegistration(@PathVariable Long courseRegistrationId) {
+        return new ResponseEntity<>(courseRegistrationService.checkByCourseRegistrationId(courseRegistrationId),HttpStatus.OK);
```
```suggestion
        return ResponseEntity.ok()
            .body(courseRegistrationService.checkByCourseRegistrationId(courseRegistrationId));
```

TwooTwoo replied:
이와 같은 방식을 '빌더 스타일' 이라고 한다는 것을 알게 되었습니다. 모든 컨트롤러 객체에 일관성 있게 적용시켜 보았습니다. 확실히 코드가 간결해지고 가독성이 높아지는 것 같습니다. 앞으로 컨트롤러 클래스를 작성할 때 참고하도록 하겠습니다!

hangillee commented on KIMTAEWOO/GDGWeek4AssignmentKTW/src/main/java/org/example/gdgweek4assignmentktw/domain/CourseRegistration.java:
```diff
@@ -0,0 +1,55 @@
+
+@Entity
+@Getter
+@NoArgsConstructor
```
```suggestion
@NoArgsConstructor(accessLevel = AccessLevel.PROTECTED)
```

TwooTwoo replied:
수정해주신 코드를 보고 왜 수정해 주셨을지 고민해 보고, 여러가지를 찾아보았습니다.
1. 혹시 자바빈규약에 위배된 걸까?
자바빈 규약은 단순히 기본 생성자를 반드시 가지고 있어야 한다는 것이므로 이건 아니라고 판단했습니다.
2. 기왕 제한할꺼 private으로 하면 되지, 왜 protected 였을까?
여기까지 생각하고 블로그 검색을 통해 답을 찾았습니다. JPA에서 엔티티 매핑 방식을 LAZY Loading 으로 했을 때 우선 프록시 객체를 만들어 놓고 실제로 접근하는 시점에 프록시 객체에 정보를 주입하는 방식으로 작동합니다. 그런데 엔티티의 프록시는 원본 객체를 상속해서 만듭니다. private 으로 선언하면 부모 객체의 생성자 super 을 호출할 수 없어 결론적으로 프록시 객체를 만들 수 없으므로 protected 여야 합니다.

- reference
https://shorturl.at/3XTLy

아직 지식의 깊이가 얕지만, 제가 정말 원하던 파고들 수 있는 코드 리뷰였다고 생각합니다. 저의 성장에 많은 도움이 되었습니다. 정말 감사합니다! :)

TwooTwoo replied:
그리고
(accessLevel = AccessLevel.PROTECTED) 이 아니라
(access = AccessLevel.PROTECTED) 으로 확인했습니다.
알려주셔서 다시 한 번 감사합니다!

hangillee replied:
훌륭하네요! 프록시까지 학습하실 줄은 몰랐습니다. 더 깊이 학습하고 싶으시다면 CGLIB의 프록시 객체 생성 방식에 대해서 알아보면 더 좋을 것 같네요.

TwooTwoo replied:
넵 꼭 시간 내서 더 알아보도록 하겠습니다. 감사합니다!

hangillee commented on KIMTAEWOO/GDGWeek4AssignmentKTW/src/main/java/org/example/gdgweek4assignmentktw/domain/Student.java:
```diff
@@ -0,0 +1,69 @@
+    cascade : 부모 엔티티 상태 변화시 자식 엔티티 어떻게 할지 설정
+    orphanRemoval : 고아 엔티티 삭제 여부
+     */
+    @OneToMany(mappedBy = "student", fetch = FetchType.LAZY
+            , cascade = CascadeType.ALL, orphanRemoval = true)
+    private List<CourseRegistration> registrations = new ArrayList<>();
```
양방향 연관관계 매핑의 단점은 무엇이 있을까요?

TwooTwoo replied:
찾아보니 정말 많은 문제가 일어날 수 있네요.
대표적으로는 일반적인 복잡성 증가 문제가 있습니다.
그 외에도 많지만 하나를 꼽자면, 주인이 아닌 쪽은 읽기 전용이라는 점이 있습니다.
하지만 테스트 코드에서 리스트에서 제거하는 과정을 통해 연관관계의 주인을 삭제하는 것처럼 보이는 코드를 작성할 수 있고, 이는 실제로 테스트도 통과합니다. **하지만 실제로 DB에는 아무런 영향을 미치지 않습니다.** 
인상깊은 내용이었습니다. 감사합니다.

hangillee commented on KIMTAEWOO/GDGWeek4AssignmentKTW/src/main/java/org/example/gdgweek4assignmentktw/dto/courseRegistration/response/CourseRegistrationListResponseDto.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.Collections;
+import java.util.List;
+
+import static com.fasterxml.jackson.databind.type.LogicalType.Collection;
+@Builder
```
여기 사이에는 개행이 있어야 할 것 같네요..

TwooTwoo replied:
인텔리제이에서 import 자동 접기가 켜져있어서 미처 확인하지 못했네요. 앞으로 신경쓰도록 하겠습니다.

hangillee commented on KIMTAEWOO/GDGWeek4AssignmentKTW/src/main/java/org/example/gdgweek4assignmentktw/dto/student/response/StudentInfoResponseDto.java:
```diff
@@ -0,0 +1,34 @@
+import lombok.Getter;
+import org.example.gdgweek4assignmentktw.domain.Student;
+
+import java.util.Date;
+
+
+@Builder
```
개행도 컨벤션의 일종입니다. 무의미한 2줄 개행은 코드의 통일성이 깨져보입니다.

TwooTwoo replied:
수정하겠습니다!

hangillee commented on KIMTAEWOO/GDGWeek4AssignmentKTW/src/main/java/org/example/gdgweek4assignmentktw/exception/CourseAlreadyExistsException.java:
커스텀 예외 활용 좋네요!

dongkyun0713 commented on KIMTAEWOO/GDGWeek4AssignmentKTW/src/main/java/org/example/gdgweek4assignmentktw/domain/CourseRegistration.java:
```diff
@@ -0,0 +1,55 @@
+모든 요소를 받는 package-private 생성자가 자동으로 생셩되며
+이 생성자에 @Builder 어노테이션을 붙인 것과 동일하게 동작한다.
+ */
+@Builder
```
클래스에 `@Builder`를 붙이는 것과 생성자에 `@Builder`붙이는 것은 무슨 차이가 있을까요?

TwooTwoo replied:
질문을 통해 파고드는 과정에서 많은 것을 알게 되었습니다.
클래스에 @Builder 을 붙이게 되면 객체 생성시 노출되지 말아야 할 매개변수도 노출될 수 있음을 깨달았습니다.
또 빌더가 점층적 생산자 패턴 -> 자바빈 패턴 -> 빌더 패턴의 과정을 통해 단점을 보완하며 만들어졌다는 것도 알게 되었습니다. 
많은 도움이 되었습니다. 감사합니다!

dongkyun0713 commented on KIMTAEWOO/GDGWeek4AssignmentKTW/src/main/java/org/example/gdgweek4assignmentktw/domain/CourseRegistration.java:
```diff
@@ -0,0 +1,55 @@
+    private LocalDateTime registrationTime;
+
+    // 정적 팩토리 메서드
+    public static CourseRegistration createWithCourseAndStudent(Course course, Student student) {
+        CourseRegistration courseRegistration = new CourseRegistration();
+        courseRegistration.course = course;
+        courseRegistration.student = student;
+        return courseRegistration;
+    }
```
빌더 패턴을 사용할 수 있을 것 같아요!

dongkyun0713 reviewed (commented):
예외처리도 꼼꼼히 하시고, 새로운 것도 많이 시도 하셨네요~! 열심히 공부하신 게 느껴져서 좋네요! 과제하시느라 수고하셨습니다! 

> 정적 팩토리 메서드에 대해 많이 찾아 보았습니다. of 와 from 등의 작명 규칙을 따를 수도 있고, 필요에 따라 적절한 이름을 작명할 수도 있는 것으로 알고 있습니다. 실무에서는 이 두가지 방식 중 어떤 것을 주로 사용하는지 궁금합니다.

실무는 저도 안 겪어봐서 모르겠지만, 보통 매개변수가 1개이면 `from` 여러개이면 `of`가 자주 사용됩니다. `from`과 `of` 모두 `어떤 것으로부터` 라는 뜻을 가지고 있죠. 예를 들어 `StudentInfoResponseDto.of(data)`이면 `date`로부터 `StudentInfoResponseDto`를 생성한다는 의미로 해석하게 되는거죠.

TwooTwoo committed:
Refactor : 전반적인 코드 리팩토링
1. POSIX New Line 적용
2. 컨트롤러 객체에 빌더 스타일 적용으로 가독성 향상
3. 보안성을 위해 엔티티 객체들의 @NoArgsConstructor 어노테이션에 (access = AccessLevel.PROTECTED) 속성 추가
4. 그외 개행, import문 정리 등 코드 스타일 개선
